### PR TITLE
IdentityProvider input parameter; refactor Attributes

### PR DESCRIPTION
**Pull request by aajtodd**
This PR was originally meant to be split in two but I messed that up somewhere along the way and it's not too big as to be cumbersome to review together.

## Issue \#

n/a

## Description of changes

* Refactor `IdentityProvider` (and subsequently `CredentialsProvider`) to take a parameter to allow for future extension (SRA requirement). 
* Refactor `Attributes` into immutable and mutable versions (`MutableAttributes`). Most existing uses of `Attributes` were updated to be `MutableAttributes` without worry or care if we could do better yet (most of our usage is from `ExecutionContext` which will be hard to make immutable). 
    * Fix endpoint generation to make use of new APIs
* Generate `authSchemes` parameter for service client configuration

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/util/RuntimeConfigProperty.kt [hunk 1]
```diff
@@ -164,4 +164,15 @@ object RuntimeConfigProperty {
             later than any added automatically by the SDK.
         """.trimIndent()
     }
+
+    val HttpAuthSchemes = ConfigProperty {
+        name = "authSchemes"
+        symbol = KotlinTypes.Collections.list(RuntimeTypes.Auth.HttpAuth.HttpAuthScheme, default = "emptyList()")
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/util/RuntimeConfigProperty.kt:170]: Also, I assume the true defaulted list is coming in follow-on?

**Review by lucix-aws — APPROVED** [review]

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/util/RuntimeConfigProperty.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (aajtodd) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/util/RuntimeConfigProperty.kt:170]: > Don't we want this to be mutable in builderland?

I went back and forth on this and came to the conclusion maybe we don't want multiple ways to do this (append vs supply new list). I opted for users needing to set the list explicitly `authSchemes = listOf(scheme1, scheme2)` vs `authSchems.add(scheme1)`. 

----

> Also, I assume the true defaulted list is coming in follow-on?

It's already in this PR. This list is only to configure new custom auth schemes OR override default ones. We build up the map of configured auth schemes in the generated client ([here](https://github.com/awslabs/smithy-kotlin/blob/13c6bf9ba5971752cd2e491ad4702bd7abd5fd73/codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGenerator.kt#L88)) by using the user provided ones if already set or computing a default if missing.

Example:
```kotlin
    private val configuredAuthSchemes = with(config.authSchemes.associateBy(HttpAuthScheme::schemeId).toMutableMap()){
        getOrPut(AuthSchemeId.AwsSigV4){
            SigV4AuthScheme(DefaultAwsSigner, "sts")
        }
        getOrPut(AuthSchemeId.Anonymous){
            AnonymousAuthScheme
        }
        toMap()
    }
```

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/Attributes.kt [hunk 2]
```diff
@@ -65,39 +63,44 @@ public operator fun <T : Any> Attributes.get(key: AttributeKey<T>): T = getOrNul
...
-public fun Attributes.merge(other: Attributes) {
+public fun MutableAttributes.merge(other: Attributes) {
     other.keys.forEach {
         @Suppress("UNCHECKED_CAST")
         set(it as AttributeKey<Any>, other[it])
     }
 }
 
-private class AttributesImpl : Attributes {
+private class AttributesImpl constructor(seed: Attributes) : MutableAttributes {
```
Comment (lauzadis) [inline runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/Attributes.kt:97]: should this be named `MutableAttributesImpl`?

**Review by lauzadis — APPROVED** [review]

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/Attributes.kt (hunk 2 above)
Comment (aajtodd) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/Attributes.kt:97]: I don't know if it matters since it's `private`

**Pull request merged**